packages/@aws-cdk/aws-lambda/lib/code.ts

```diff
@@ -248,6 +248,8 @@ export class AssetCode extends Code {
         path: this.path,
         ...this.options,
       });
+    } else if (cdk.Stack.of(this.asset) !== cdk.Stack.of(scope)) {
+      throw new Error(`Asset is already associated with another stack '${cdk.Stack.of(this.asset).stackName}'.`);
```

Comment:
When writing an error message, try to also suggest what the user needs to do to fix it (even if it's super obvious to you, it might not be to the user seeing the error). Add something like:

"Create a new Code instance for every stack."

Provisional approval but I hope you'll consider this.